src/dwarf_parser.cpp

```diff
@@ -1,250 +1,213 @@
 #include "dwarf_parser.h"
 
-#ifdef HAVE_LIBDW
+#ifdef HAVE_LIBLLDB
 
 #include "bpftrace.h"
 #include "log.h"
+#include "types.h"
 
-#include <dwarf.h>
-#include <elfutils/libdw.h>
+#include <atomic>
+#include <lldb/API/LLDB.h>
+#include <llvm/Config/llvm-config.h>
+#include <memory>
+#include <string>
 
 namespace bpftrace {
 
-struct FuncInfo {
-  std::string name;
-  Dwarf_Die die;
-};
+std::atomic<size_t> Dwarf::instance_count = 0;
 
-Dwarf::Dwarf(BPFtrace *bpftrace, const std::string &file_path)
-    : bpftrace_(bpftrace), file_path_(file_path)
+Dwarf::Dwarf(BPFtrace *bpftrace, std::string file_path)
+    : bpftrace_(bpftrace), file_path_(std::move(file_path))
 {
-  callbacks.find_debuginfo = dwfl_standard_find_debuginfo;
-  callbacks.section_address = dwfl_offline_section_address;
-  callbacks.debuginfo_path = NULL;
-  dwfl = dwfl_begin(&callbacks);
-  dwfl_report_offline(dwfl, file_path.c_str(), file_path.c_str(), -1);
-  dwfl_report_end(dwfl, NULL, NULL);
-}
+  if (instance_count++ == 0)
+    lldb::SBDebugger::Initialize();
 
-std::unique_ptr<Dwarf> Dwarf::GetFromBinary(BPFtrace *bpftrace,
-                                            const std::string &file_path)
-{
-  std::unique_ptr<Dwarf> dwarf(new Dwarf(bpftrace, file_path));
-  Dwarf_Addr bias;
-  if (dwfl_nextcu(dwarf->dwfl, NULL, &bias) == NULL)
-    return nullptr;
+  debugger_ = lldb::SBDebugger::Create(false);
 
-  return dwarf;
+  lldb::SBError error;
+  target_ = debugger_.CreateTarget(
+      file_path_.c_str(), nullptr, nullptr, false, error);
+  if (!error.Success() || !target_.IsValid()) {
+    throw error;
+  }
 }
 
 Dwarf::~Dwarf()
 {
-  dwfl_end(dwfl);
+  if (--instance_count == 0)
+    lldb::SBDebugger::Terminate();
 }
 
-static int get_func_die_cb(Dwarf_Die *func_die, void *arg)
+std::unique_ptr<Dwarf> Dwarf::GetFromBinary(BPFtrace *bpftrace,
+                                            std::string file_path)
 {
-  auto *func_info = static_cast<struct FuncInfo *>(arg);
-  if (dwarf_hasattr(func_die, DW_AT_name) &&
-      dwarf_diename(func_die) == func_info->name) {
-    func_info->die = *func_die;
-    return DWARF_CB_ABORT;
+  try {
+    return std::unique_ptr<Dwarf>(new Dwarf(bpftrace, std::move(file_path)));
```

Comment:
Unfortunately, only Clang handles the friendship for `std::make_unique` and GCC doesn't, see the [CI run](https://github.com/bpftrace/bpftrace/actions/runs/8330811720/job/22796307638?pr=3042#step:6:1501).
I'll revert to the old way and double check we don't leak a `Dwarf` object in case of exception.